File: app/src/main/assets/www/com.emmoco.core/SerialPacket.js
if (typeof define !== 'function') { var define = require('amdefine')(module); }

define(function (require) {

	'use strict';
	
	/* -------- IMPORTS -------- */
	var Utils = require('com.emmoco.core/Utils');

	var DATA_SIZE = 240;
	var HDR_SIZE = 4;
	
	var PktKind = Utils.EnumT(
		'NOP',
		'FETCH',
		'FETCH_DONE',
		'STORE',
		'STORE_DONE',
		'INDICATOR',
		'CONNECT',
		'DISCONNECT',
		'ECHO',
		'PAIRING',
		'PAIRING_DONE',
		'OFFLINE',
		'ACCEPT',
		'START',
		'ACTIVE_PARAMS',
		'SCAN',
		'SCAN_DONE',
		'BEACON'
	);
	
	function PktHdr() {
	}

	PktHdr.prototype.toString = function() {
		return Utils.sprintf("<size: %d, kind: %s, resid: %d>", this.size, PktKind[this.kind], this.resId);
	};

	function create(buf) {
		return init.call(new SerialPacket(), buf);
	}

	function addHdr(kind, resId, chan) {
		this.mHdr = false;
		addInt8.call(this, HDR_SIZE);
		addInt8.call(this, kind);
		addInt8.call(this, resId ? resId : 0);
		addInt8.call(this, chan ? chan : 0);
		this.mHdr = true;
	}
	
	function addInt8(d) {
		this.mBuffer[this.mCurIdx++] = (d & 0xFF) - (d < 0 ? 256 : 0);
		incSize.call(this, 1);
	}
	
	function addInt16(d) {
		this.mBuffer[this.mCurIdx++] = d & 0xFF;
		this.mBuffer[this.mCurIdx++] = ((d >> 8) & 0xFF) - (d < 0 ? 256 : 0);
		incSize.call(this, 2);
	}
	
	function addInt32(d) {
		this.mBuffer[this.mCurIdx++] = d & 0xFF;
		this.mBuffer[this.mCurIdx++] = (d >> 8) & 0xFF;
		this.mBuffer[this.mCurIdx++] = (d >> 16) & 0xFF;
		this.mBuffer[this.mCurIdx++] = ((d >> 24) & 0xFF) - (d < 0 ? 256 : 0);
		incSize.call(this, 4);
	}
	
	function alignTo(align) {
		var off = this.mCurIdx % align;
		if (off) {
			var inc = align -off;
			this.mCurIdx += inc;
			incSize.call(this, inc);
		}
	}
	
	function incSize(sz) {
		if (this.mHdr) {
			this.mBuffer[0] += sz;
		}
	}
	
	function init(buf) {
		if (buf) {
			this.mBuffer = buf;
			this.mCurIdx = buf.length;
		}
		else {
			this.mBuffer = new Uint8Array(HDR_SIZE + DATA_SIZE);
			this.mCurIdx = 0;
		}
		this.mHdr = false;
		return this;
	}
	
	function getData() {
		return this.mBuffer.subarray(HDR_SIZE, this.mBuffer[0]);
	}
	
	function getHdr() {
		var idx = this.mCurIdx;
		this.mCurIdx = 0;
		var hdr = scanHdr.call(this);
		this.mCurIdx = idx;
		return hdr;
	}
	
	function rewind() {
		this.mCurIdx = 0;
	}
	
	function scanHdr() {
		var hdr = new PktHdr();
		hdr.size = scanUns8.call(this);
		hdr.kind = scanUns8.call(this);
		hdr.resId = scanInt8.call(this);
		hdr.chan = scanUns8.call(this);
		return hdr;
	}
	
	function scanInt8() {
		var b0 = this.mBuffer[this.mCurIdx++];
		b0 = b0 < 128 ? b0 : (b0 - 256);
		return b0;
	}
	
	function scanInt16() {
		var b0 = this.mBuffer[this.mCurIdx++];
		var b1 = this.mBuffer[this.mCurIdx++];
		b1 = b1 < 128 ? b1 : (b1 - 256);
		return b0 + (b1 << 8);
	}
	
	function scanInt32() {
		var b0 = this.mBuffer[this.mCurIdx++];
		var b1 = this.mBuffer[this.mCurIdx++];
		var b2 = this.mBuffer[this.mCurIdx++];
		var b3 = this.mBuffer[this.mCurIdx++];
		b3 = b3 < 128 ? b3 : (b3 - 256);
		return b0 + (b1 << 8) + (b2 << 16) + (b3 << 24);
	}
	
	function scanUns8() {
		var b0 = this.mBuffer[this.mCurIdx++];
		return b0 < 0 ? (b0 + 256) : b0;
	}
	
	function scanUns16() {
		var b0 = this.mBuffer[this.mCurIdx++];
		var b1 = this.mBuffer[this.mCurIdx++];
		return b0 + (b1 << 8);
	}
	
	function scanUns32() {
		var b0 = this.mBuffer[this.mCurIdx++];
		var b1 = this.mBuffer[this.mCurIdx++];
		var b2 = this.mBuffer[this.mCurIdx++];
		var b3 = this.mBuffer[this.mCurIdx++];
		var b4 = 0;
		if (b3 > 127) {
			b3 = b3 - 128;
			b4 = 0x80000000;
		}
		return b0 + (b1 << 8) + (b2 << 16) + (b3 << 24) + b4;
	}
	
	/* -------- EXPORTS -------- */
	function SerialPacket() {}
	SerialPacket.prototype = Object.defineProperties({}, {
		buffer: { get: function() { return this.mBuffer; }},
		length:	{ get: function() { return this.mCurIdx; }},
	});
	SerialPacket.DATA_SIZE = DATA_SIZE;
	SerialPacket.HDR_SIZE = HDR_SIZE;
	SerialPacket.PktKind = PktKind;
	SerialPacket.create = create;
	SerialPacket.prototype.addHdr = addHdr;
	SerialPacket.prototype.addInt8 = addInt8;
	SerialPacket.prototype.addInt16 = addInt16;
	SerialPacket.prototype.addInt32 = addInt32;
	SerialPacket.prototype.alignTo = alignTo;
	SerialPacket.prototype.getData = getData;
	SerialPacket.prototype.getHdr = getHdr;
	SerialPacket.prototype.rewind = rewind;
	SerialPacket.prototype.scanHdr = scanHdr;
	SerialPacket.prototype.scanInt8 = scanInt8;
	SerialPacket.prototype.scanInt16 = scanInt16;
	SerialPacket.prototype.scanInt32 = scanInt32;
	SerialPacket.prototype.scanUns8 = scanUns8;
	SerialPacket.prototype.scanUns16 = scanUns16;
	SerialPacket.prototype.scanUns32 = scanUns32;
	return SerialPacket;
});

File: app/src/main/java/com/emmoco/emgap/SerialPacket.java
package com.emmoco.emgap;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;

public class SerialPacket {

    static public final int SYS_CHAN = 0x80;

    static public final int CLIENT_ADMIN_RESID = 0;
    static public final int CLIENT_ADDR_RESID = 1;
    static public final int CLIENT_SCAN_RESID = 2;
    static public final int CLIENT_SCANDUR_RESID = 3;
    static public final int CLIENT_SCANINFO_RESID = 4;

    public static enum Kind {
        NOP,
        FETCH,
        FETCH_DONE,
        STORE,
        STORE_DONE,
        INDICATOR,
        CONNECT,
        DISCONNECT,
        ECHO,
        PAIRING,
        PAIRING_DONE,
        OFFLINE,
        ACCEPT,
		START,
		ACTIVE_PARAMS,
		SCAN,
		SCAN_DONE,
		BEACON,
    }

    public static class Header {
        public int size;
        public Kind kind;
        public int resId;
        public int chan;
        public String toString() {
            return String.format("<%d,%s,%d,%02x>", size, kind, resId, chan);
        }
    }

    public static final int HDR_SIZE = 4;
    public static final int MAX_DATA_SIZE = 240;

    private final byte[] mBuffer;
    private int mCurIdx;
    private boolean mHdrFlag;

    public SerialPacket() {
        mBuffer = new byte[MAX_DATA_SIZE + HDR_SIZE];
    }

    public void addHeader(Header hdr) {
        rewind();
        addInt8(hdr.size + HDR_SIZE);
        addInt8(hdr.kind.ordinal());
        addInt8(hdr.resId);
        addInt8(hdr.chan);
    }

    public void addHeader(Kind kind) {
    	addHeader(kind, 0, 0);
    }
    
    public void addHeader(Kind kind, int resId) {
    	addHeader(kind, resId, 0);
    }
    
    public void addHeader(Kind kind, int resId, int chan) {
    	mHdrFlag = false;
        addInt8(HDR_SIZE);
        addInt8(kind.ordinal());
        addInt8(resId);
        addInt8(chan);
    	mHdrFlag = true;
    }
    
    public void addInt8(int d) {
        mBuffer[mCurIdx++] = (byte) (d & 0xFF);
        incSize(1);
    }

    public void addInt16(int d) {
        mBuffer[mCurIdx++] = (byte) (d & 0xFF);
        mBuffer[mCurIdx++] = (byte) ((d >> 8) & 0xFF);
        incSize(2);
    }

    public void addInt32(long d) {
        mBuffer[mCurIdx++] = (byte) (d & 0xFF);
        mBuffer[mCurIdx++] = (byte) ((d >> 8) & 0xFF);
        mBuffer[mCurIdx++] = (byte) ((d >> 16) & 0xFF);
        mBuffer[mCurIdx++] = (byte) ((d >> 24) & 0xFF);
        incSize(4);
    }

    public void alignTo(int align) {
        int off = mCurIdx % align;
        if (off != 0) {
        	int inc = align - off;
            mCurIdx += inc;
            incSize(inc);
        }
    }

    public void clear() {
        for (int i = 0; i < mBuffer.length; i++) {
            mBuffer[i] = 0;
        }
    }
    
    public byte[] getBytes() {
    	return Arrays.copyOf(mBuffer, mCurIdx);
    }
    
    public byte[] getBuffer() {
        return mBuffer;
    }
    
	public byte[] getData() {
    	return Arrays.copyOfRange(mBuffer, HDR_SIZE, mBuffer[0] & 0xFF);
	}
	
    public int getSize() {
        return mCurIdx;
    }
    
    private void incSize(int sz) {
    	if (mHdrFlag) {
    		mBuffer[0] += sz;
    	}
    }
    
    public void recv(InputStream in) throws Exception {
        rewind();
        clear();
        int size = in.read();
        if (size < 0) {
            return;
        }
        mBuffer[mCurIdx++] = (byte) size;
        while (mCurIdx < size) {
            int b = in.read();
            mBuffer[mCurIdx++] = (byte) b;
        }
    }

    public void rewind() {
        mCurIdx = 0;
    }

    public void scanHeader(Header hdr) {
        rewind();
        hdr.size = scanUns8();
        hdr.kind = Kind.values()[scanUns8()];
        hdr.resId = scanInt8();
        hdr.chan = scanUns8();
    }

    public int scanInt8() {
        return mBuffer[mCurIdx++];
    }

    public int scanInt16() {
        int b0 = mBuffer[mCurIdx++];
        b0 = b0 < 0 ? (b0 + 256) : b0;
        int b1 = mBuffer[mCurIdx++];
        return b0 + (b1 << 8);
    }

    public long scanInt32() {
        int b0 = mBuffer[mCurIdx++];
        b0 = b0 < 0 ? (b0 + 256) : b0;
        int b1 = mBuffer[mCurIdx++];
        b1 = b1 < 0 ? (b1 + 256) : b1;
        int b2 = mBuffer[mCurIdx++];
        b2 = b2 < 0 ? (b2 + 256) : b2;
        int b3 = mBuffer[mCurIdx++];
        return b0 + (b1 << 8) + (b2 << 16) + (b3 << 24);
    }

    public int scanUns8() {
        int b0 = mBuffer[mCurIdx++];
        return b0 < 0 ? (b0 + 256) : b0;
    }

    public int scanUns16() {
        int b0 = mBuffer[mCurIdx++];
        b0 = b0 < 0 ? (b0 + 256) : b0;
        int b1 = mBuffer[mCurIdx++];
        b1 = b1 < 0 ? (b1 + 256) : b1;
        return b0 + (b1 << 8);
    }

    public long scanUns32() {
        long b0 = mBuffer[mCurIdx++];
        b0 = b0 < 0 ? (b0 + 256) : b0;
        long b1 = mBuffer[mCurIdx++];
        b1 = b1 < 0 ? (b1 + 256) : b1;
        long b2 = mBuffer[mCurIdx++];
        b2 = b2 < 0 ? (b2 + 256) : b2;
        long b3 = mBuffer[mCurIdx++];
        b3 = b3 < 0 ? (b3 + 256) : b3;
        return b0 + (b1 << 8) + (b2 << 16) + (b3 << 24);
    }

    public void send(OutputStream out) throws Exception {
        out.write(mBuffer, 0, mCurIdx);
        out.flush();
        rewind();
    }
}
